factor/totp/classes/factor.php

```diff
@@ -90,21 +72,63 @@ public function define_add_factor_form_definition($mform) {
         return $mform;
     }
 
-    public function define_login_form($mform) {
-        global $OUTPUT, $USER;
-        $userfactors = $this->get_user_factors($USER->id);
+    public function add_factor_form_definition_after_data($mform) {
+        global $OUTPUT;
+        $secretfield = $mform->getElement('secret');
 
-        foreach ($userfactors as $userfactor) {
-            if ($userfactor->disabled != 1) {
-                $mform->addElement('text', 'verificationcode', get_string('verificationcode', 'factor_totp'));
-                $mform->addRule('verificationcode', get_string('required'), 'required', null, 'client');
-                $mform->setType("verificationcode", PARAM_ALPHANUM);
-            }
+        //if (!empty($secretfield)) {
+        $secret = $secretfield->getValue();
+        $qrcode = $this->draw_qrcode($secret);
+
+        $mform->addElement('html', $OUTPUT->heading('Scan QR-Code or enter a key to your Google Authenticator:', 5));
+        $mform->addElement('html', $OUTPUT->heading($this->get_secret_length().'-digit key: '.$secret, 5));
+        $mform->addElement('html', $qrcode);
+        $mform->addElement('html', $OUTPUT->box(''));
+        //}
+
+        return $mform;
+    }
+
+    public function add_factor_form_validation($data) {
+        $errors = array();
+
+        $totp = TOTP::create($data['secret']);
+        if ($data['verificationcode'] != $totp->now()) {
+            $errors['verificationcode'] = get_string('error:wrongverification', 'factor_totp');
+        }
+
+        return $errors;
+    }
+
+    public function login_form_definition($mform) {
+        $userfactors = $this->get_enabled_user_factors();
+
+        if (count($userfactors) > 0) {
+            $mform->addElement('text', 'verificationcode', get_string('verificationcode', 'factor_totp'));
+            $mform->addRule('verificationcode', get_string('required'), 'required', null, 'client');
+            $mform->setType("verificationcode", PARAM_ALPHANUM);
         }
 
         return $mform;
     }
 
+    public function login_form_validation($data) {
+        $factors = $this->get_enabled_user_factors();
+
+        foreach ($factors as $factor) {
+            if ($factor) {
+                $secret = $factor->secret;
+                $hotp = TOTP::create($secret);
+                $otp = $hotp->now();
+
+                if ($data['verificationcode'] != $otp) {
+                    return array('verificationcode' => 'Wrong verification code');
+                }
+            }
+        }
+        return array();
+    }
+
     public function get_secret_length() {
         $length = get_config('factor_totp', 'secret_length');
```

Comment:
I would actually remove this setting and hard code it to something very secure which is known to work everywhere. The less settings the better

Reply:
Will be done in further PR